binding.c

```diff
@@ -92,40 +98,93 @@ parse_address (struct sockaddr *name, char *ip, size_t size, int *port, int *fam
   }
 }
 
+/**
+ * When a native function is executing on uv event;
+ * (cannot be surrounded by try-catch block)
+ *
+ * redirect pending exceptions to global uncaught handler
+ * to avoid process abort (given an uncaught handler is installed).
+ */
+static inline void
+redirect_to_uncaught (js_env_t *env, int err) {
+  if (err == 0) return;
+  assert(err == js_pending_exception || err == js_uncaught_exception);
+
+  bool is_pending;
+  err = js_is_exception_pending(env, &is_pending);
+  assert(err == 0);
+
+  if (is_pending) {
+    js_value_t *fatal_exception;
+    err = js_get_and_clear_last_exception(env, &fatal_exception);
+    assert(err == 0);
+
+    err = js_fatal_exception(env, fatal_exception);
+    assert(err == 0);
+  }
+}
+
+static inline int
+get_buffer_info (js_env_t *env, js_value_t *buffer, void **data, size_t *len) {
+#ifdef CHECKED_GET_BUFFER_INFO
+  int err;
+  bool is_typedarray;
+  err = js_is_typedarray(env, buffer, &is_typedarray);
+  if (err) return err;
+  if (is_typedarray) return js_get_typedarray_info(env, buffer, NULL, data, len, NULL, NULL);
+
+  bool is_arraybuffer;
+  err = js_is_arraybuffer(env, buffer, &is_arraybuffer);
+  if (err) return err;
+  if (is_arraybuffer) return js_get_arraybuffer_info(env, buffer, data, len);
+
+  return -1;
+#else
+  // all buffers allocated through b4a are typed-arrays
+  return js_get_typedarray_info(env, buffer, NULL, data, len, NULL, NULL);
+#endif
+}
```

Comment:
We should make sure that we only ever pass typed arrays back to native to avoid potentially calling into the engine 4 times instead of just once.

Reply:
I removed the checks; we are of course only passing typed arrays.
But i kept the private `get_buffer_info` function as a reminder to move some allocations to native in the future.

Reply:
I'd remove it and just use `js_get_typedarray_info()`. Might be more appropriate to add some explicit `// TODO` comments for the cases where we can allocate on the native side instead.

Reply:
helper function removed.

Todos not added, but it's a good candidate for different PR.
(Not touching allocs right now, because `index.js` and `test.js` are my source of truth.)